retrofit2/src/main/java/com/linecorp/armeria/client/retrofit2/RetrofitMeterIdPrefixFunctionBuilder.java

```diff
@@ -33,11 +33,18 @@
     private String serviceTagName;
     @Nullable
     private String defaultServiceName;
+    @Nullable
+    private Class<?> serviceClass;
 
     RetrofitMeterIdPrefixFunctionBuilder(String name) {
         this.name = name;
     }
 
+    RetrofitMeterIdPrefixFunctionBuilder(String name, Class<?> serviceClass) {
+        this.name = name;
+        this.serviceClass = serviceClass;
```

Comment:
We should make sure that both `serviceClass` and `defaultServiceName` are not set together.

Reply:
Maybe I'm wrong, but I don't think it would affect usability:
If user use this builder then it's already clear that he want to use `serviceClass` but he still able to modify `serviceTag` with `withServiceTag` method.

Maybe comment on `withServiceTag` will help on this to clarify?

Reply:
Isn't `defaultServiceName` unused at all when `serviceClass` is used? How about:

- throwing an `IllegalStateException("defaultServiceName cannot be set for a builder created with serviceClass.")` for `withServiceTag(String, String)`,
- and adding `withServiceTag(String serviceTagName)` which works for both builder types.
  - If created without serviceClass, simply delegate to `withServiceTag(serviceTagName, "UNKNOWN")`

..?

Reply:
Why not to split this method into 2 and mark original `withServiceTag(String, String)` as deprecated?
- `withServiceTag(String serviceTag)`
- `withServiceName(String serviceName)` <-- I was curious why we need to provide `defaultServiceName` that used only in cases when `Invocation` not found (see [code](https://github.com/line/armeria/blob/master/retrofit2/src/main/java/com/linecorp/armeria/client/retrofit2/RetrofitMeterIdPrefixFunction.java#L106)). Let's just provide ability to use client provided service name that would be used for all cases and don't need to take name from the `Invocation` log then

Reply:
So this code from the link above:
```java
        if (invocation != null) {
            if (serviceTagName != null) {
                tagListBuilder.add(Tag.of(serviceTagName,
                                          invocation.method().getDeclaringClass().getSimpleName()));
            }
            tagListBuilder.add(Tag.of("method", invocation.method().getName()));
        } else {
            if (serviceTagName != null) {
                assert defaultServiceName != null;
                tagListBuilder.add(Tag.of(serviceTagName, defaultServiceName));
            }
            tagListBuilder.add(Tag.of("method", log.method().name()));
        }
```
Would be simplified to:
```
        if (serviceTagName != null) {
               tagListBuilder.add(Tag.of(serviceTagName, serviceName));
        }
        if (invocation != null) {
            tagListBuilder.add(Tag.of("method", invocation.method().getName()));
        } else {
            tagListBuilder.add(Tag.of("method", log.method().name()));
        }
```

Reply:
Unfortunately we can't make this ^ change now as it will break current metrics (as it may happen that provided serviceName would be different), but I think it should be way to go in the future

Reply:
> Why not to split this method into 2 and mark original `withServiceTag(String, String)` as deprecated?

Sounds good to me!

Reply:
How about deprecating the current `withServiceTag` and adding Two methods:
- `withServiceName(String serviceTagName, String defaultServiceName)`
  - We should mention when the `defaultServiceName` is used in Javadoc
    - `@param defaultServiceName the default value of the tag, e.g. {@code "myService"}. It's only used when {@link InvocationUtil#getInvocation(RequestLog)} returns {@code null}.`
- `withServiceCalss(String serviceTagName, Class<?> serviceClass)`

Reply:
I think it's better to separate into `withServiceName(serviceName)` and `withServiceTag(serviceTagName)` methods as we may want to override service name for the case when class used. Also we may just introduce meaningful defaults to reduce code for users.

Reply:
> we may want to override service name for the case when class used

Ah, yes I missed that point. 😄 

So, now we can set the property one by one, how about removing the prefix `with`?
Also, What do you think adding a setter for the serviceClass so that a user can do:
```
RetrofitMeterIdPrefixFunction.of(name)
                             .serviceTag(serviceTagName)
                             .serviceClass(serviceClass) // I'm not sure if this is a good name though. ;)
                             .build()
```